bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LedgerDirsMonitor.java

```diff
@@ -0,0 +1,191 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package org.apache.bookkeeper.bookie;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+
+import org.apache.bookkeeper.bookie.LedgerDirsManager.LedgerDirsListener;
+import org.apache.bookkeeper.bookie.LedgerDirsManager.NoWritableLedgerDirException;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.util.DiskChecker;
+import org.apache.bookkeeper.util.DiskChecker.DiskErrorException;
+import org.apache.bookkeeper.util.DiskChecker.DiskOutOfSpaceException;
+import org.apache.bookkeeper.util.DiskChecker.DiskWarnThresholdException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Thread to monitor the disk space periodically.
+ */
+class LedgerDirsMonitor extends BookieThread {
+    private final static Logger LOG = LoggerFactory.getLogger(LedgerDirsMonitor.class);
+    
+    private final int interval;
+    private final ServerConfiguration conf;
+    private final ConcurrentMap<File, Float> diskUsages;
+    private final DiskChecker diskChecker;
+    private final LedgerDirsManager ldm;
+
+    public LedgerDirsMonitor(final ServerConfiguration conf, 
+            final DiskChecker diskChecker,
+            final LedgerDirsManager ldm) {
+        super("LedgerDirsMonitorThread");
+        this.interval = conf.getDiskCheckInterval();
+        this.conf = conf;
+        this.diskChecker = diskChecker;
+        this.diskUsages = ldm.getDiskUsages();
+        this.ldm = ldm;
+    }
+
+    @Override
+    public void run() {
+        while (true) {
+            try {
+                List<File> writableDirs = ldm.getWritableLedgerDirs();
+                // Check all writable dirs disk space usage.
+                for (File dir : writableDirs) {
+                    try {
+                        diskUsages.put(dir, diskChecker.checkDir(dir));
```

Comment:
@revans2 It's been over 6 months since I worked on this so may have forgotten some details.
Main point of separation of LedgerDirsMonitor and LedgerDirsManager was to break circular dependency between these two and reduce coupling. Functionally everything is same as it was minus manager having reference to monitor and dealing with its lifecycle (that is done explicitly now).
Circular dependency between two components is a strong smell of bad design (from complicating one's understanding of the code to complicating test mocks etc.), I hope we agree on that.
I left callbacks as they were (monitor calling manager's callbacks directly instead of notifying it and letting it call its callback). I agree that this can be refactored further but I did not want to get ratholed into perfecting this over other task on hands.
There is no reason for manager to handle monitor's state so that dependency is dropped. IMO it simplifies one's understanding of what's going on with that code.
I do not see problem with class being in separate file.  if you do please elaborate.